{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "sys.path[0] = '../'\n",
    "\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)\n",
    "\n",
    "from bin.build_gfedb import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '../../data/csv/'\n",
    "csv_files = os.listdir(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = {}\n",
    "\n",
    "for file in csv_files:\n",
    "    dfs[file] = pd.read_csv(path + file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['seq_edges.csv', 'sequence_nodes.csv', 'cds_edges.csv', 'gfe_edges.csv', 'cds_nodes.csv', 'group_edges.csv', 'allele_nodes.csv'])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alleleId:ID(ALLELE)</th>\n",
       "      <th>name</th>\n",
       "      <th>alleletype:LABEL</th>\n",
       "      <th>locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>HLA-Aw2-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-4</td>\n",
       "      <td>GFE</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>HLA-A*01:01:01:01</td>\n",
       "      <td>IMGT_HLA</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>HLA-A*01:01:01G</td>\n",
       "      <td>G</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>HLA-A*01:01g</td>\n",
       "      <td>lg</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>lgx</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>A*01:01</td>\n",
       "      <td>2nd_FIELD</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>HLA-Aw3-1-1-1-2-1-1-1-1-1-1-1-1-1-1-1-3</td>\n",
       "      <td>GFE</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>HLA-A*01:01:01:02N</td>\n",
       "      <td>IMGT_HLA</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>A*01:01N</td>\n",
       "      <td>2nd_FIELD</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>HLA-Aw2-1-2-1-1-1-1-1-1-1-1-1-1-1-1-1-19</td>\n",
       "      <td>GFE</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>HLA-A*01:01:01:03</td>\n",
       "      <td>IMGT_HLA</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>HLA-Aw46-1-1-1-1-1-90-1-1-1-1-1-1-1-1-1-18</td>\n",
       "      <td>GFE</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>HLA-A*01:01:01:04</td>\n",
       "      <td>IMGT_HLA</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>14</td>\n",
       "      <td>HLA-Aw333-1-1-1-280-1-1-1-1-1-1-1-1-1-1-1-4</td>\n",
       "      <td>GFE</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>15</td>\n",
       "      <td>HLA-A*01:01:01:05</td>\n",
       "      <td>IMGT_HLA</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "      <td>HLA-Aw333-1-1-1-1-1-1-1-1-1-219-1-1-1-1-1-6</td>\n",
       "      <td>GFE</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>17</td>\n",
       "      <td>HLA-A*01:01:01:06</td>\n",
       "      <td>IMGT_HLA</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>18</td>\n",
       "      <td>HLA-Aw333-1-1-1-279-1-1-1-1-1-1-1-1-1-1-1-4</td>\n",
       "      <td>GFE</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>19</td>\n",
       "      <td>HLA-A*01:01:01:07</td>\n",
       "      <td>IMGT_HLA</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>20</td>\n",
       "      <td>HLA-Aw363-1-1-1-1-1-308-1-1-1-1-1-1-1-1-1-119</td>\n",
       "      <td>GFE</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    alleleId:ID(ALLELE)                                           name alleletype:LABEL  locus\n",
       "0                     1        HLA-Aw2-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-4              GFE  HLA-A\n",
       "1                     2                              HLA-A*01:01:01:01         IMGT_HLA  HLA-A\n",
       "2                     3                                HLA-A*01:01:01G                G  HLA-A\n",
       "3                     4                                   HLA-A*01:01g               lg  HLA-A\n",
       "4                     5                                    HLA-A*01:01              lgx  HLA-A\n",
       "5                     6                                        A*01:01        2nd_FIELD  HLA-A\n",
       "6                     7        HLA-Aw3-1-1-1-2-1-1-1-1-1-1-1-1-1-1-1-3              GFE  HLA-A\n",
       "7                     8                             HLA-A*01:01:01:02N         IMGT_HLA  HLA-A\n",
       "8                     9                                       A*01:01N        2nd_FIELD  HLA-A\n",
       "9                    10       HLA-Aw2-1-2-1-1-1-1-1-1-1-1-1-1-1-1-1-19              GFE  HLA-A\n",
       "10                   11                              HLA-A*01:01:01:03         IMGT_HLA  HLA-A\n",
       "11                   12     HLA-Aw46-1-1-1-1-1-90-1-1-1-1-1-1-1-1-1-18              GFE  HLA-A\n",
       "12                   13                              HLA-A*01:01:01:04         IMGT_HLA  HLA-A\n",
       "13                   14    HLA-Aw333-1-1-1-280-1-1-1-1-1-1-1-1-1-1-1-4              GFE  HLA-A\n",
       "14                   15                              HLA-A*01:01:01:05         IMGT_HLA  HLA-A\n",
       "15                   16    HLA-Aw333-1-1-1-1-1-1-1-1-1-219-1-1-1-1-1-6              GFE  HLA-A\n",
       "16                   17                              HLA-A*01:01:01:06         IMGT_HLA  HLA-A\n",
       "17                   18    HLA-Aw333-1-1-1-279-1-1-1-1-1-1-1-1-1-1-1-4              GFE  HLA-A\n",
       "18                   19                              HLA-A*01:01:01:07         IMGT_HLA  HLA-A\n",
       "19                   20  HLA-Aw363-1-1-1-1-1-308-1-1-1-1-1-1-1-1-1-119              GFE  HLA-A"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs['allele_nodes.csv'].head(20) # HLA-A*01:01:01:01"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alleleId:ID(ALLELE)</th>\n",
       "      <th>name</th>\n",
       "      <th>alleletype:LABEL</th>\n",
       "      <th>locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>HLA-Aw2-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-4</td>\n",
       "      <td>GFE</td>\n",
       "      <td>HLA-A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   alleleId:ID(ALLELE)                                     name alleletype:LABEL  locus\n",
       "0                    1  HLA-Aw2-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-4              GFE  HLA-A"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs['allele_nodes.csv'][dfs['allele_nodes.csv']['name'] == 'HLA-Aw2-1-1-1-1-1-1-1-1-1-1-1-1-1-1-1-4']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequenceId:ID(SEQUENCE)</th>\n",
       "      <th>sequence</th>\n",
       "      <th>name</th>\n",
       "      <th>feature:LABEL</th>\n",
       "      <th>rank</th>\n",
       "      <th>length</th>\n",
       "      <th>seq:string[]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>CAGGAGCAGAGGGGTCAGGGCGAAGTCCCAGGGCCCCAGGCGTGGC...</td>\n",
       "      <td>SEQUENCE</td>\n",
       "      <td>SEQUENCE</td>\n",
       "      <td>0</td>\n",
       "      <td>3503</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>CAGGAGCAGAGGGGTCAGGGCGAAGTCCCAGGGCCCCAGGCGTGGC...</td>\n",
       "      <td>FIVE_PRIME_UTR</td>\n",
       "      <td>FEATURE</td>\n",
       "      <td>1</td>\n",
       "      <td>300</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>ATGGCCGTCATGGCGCCCCGAACCCTCCTCCTGCTACTCTCGGGGG...</td>\n",
       "      <td>EXON</td>\n",
       "      <td>FEATURE</td>\n",
       "      <td>1</td>\n",
       "      <td>73</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>GTGAGTGCGGGGTCGGGAGGGAAACCGCCTCTGCGGGGAGAAGCAA...</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>FEATURE</td>\n",
       "      <td>1</td>\n",
       "      <td>130</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...</td>\n",
       "      <td>EXON</td>\n",
       "      <td>FEATURE</td>\n",
       "      <td>2</td>\n",
       "      <td>270</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>GTGAGTGACCCCGGCCCGGGGCGCAGGTCACGACCCCTCATCCCCC...</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>FEATURE</td>\n",
       "      <td>2</td>\n",
       "      <td>241</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>GTTCTCACACCATCCAGATAATGTATGGCTGCGACGTGGGGCCGGA...</td>\n",
       "      <td>EXON</td>\n",
       "      <td>FEATURE</td>\n",
       "      <td>3</td>\n",
       "      <td>276</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>GTACCAGGGGCCACGGGGCGCCTCCCTGATCGCCTATAGATCTCCC...</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>FEATURE</td>\n",
       "      <td>3</td>\n",
       "      <td>579</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>ACCCCCCCAAGACACATATGACCCACCACCCCATCTCTGACCATGA...</td>\n",
       "      <td>EXON</td>\n",
       "      <td>FEATURE</td>\n",
       "      <td>4</td>\n",
       "      <td>276</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>GTAAGGAGGGAGATGGGGGTGTCATGTCTCTTAGGGAAAGCAGGAG...</td>\n",
       "      <td>INTRON</td>\n",
       "      <td>FEATURE</td>\n",
       "      <td>4</td>\n",
       "      <td>102</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sequenceId:ID(SEQUENCE)                                           sequence            name feature:LABEL  rank  length  seq:string[]\n",
       "0                        1  CAGGAGCAGAGGGGTCAGGGCGAAGTCCCAGGGCCCCAGGCGTGGC...        SEQUENCE      SEQUENCE     0    3503           NaN\n",
       "1                        2  CAGGAGCAGAGGGGTCAGGGCGAAGTCCCAGGGCCCCAGGCGTGGC...  FIVE_PRIME_UTR       FEATURE     1     300           NaN\n",
       "2                        3  ATGGCCGTCATGGCGCCCCGAACCCTCCTCCTGCTACTCTCGGGGG...            EXON       FEATURE     1      73           NaN\n",
       "3                        4  GTGAGTGCGGGGTCGGGAGGGAAACCGCCTCTGCGGGGAGAAGCAA...          INTRON       FEATURE     1     130           NaN\n",
       "4                        5  GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...            EXON       FEATURE     2     270           NaN\n",
       "5                        6  GTGAGTGACCCCGGCCCGGGGCGCAGGTCACGACCCCTCATCCCCC...          INTRON       FEATURE     2     241           NaN\n",
       "6                        7  GTTCTCACACCATCCAGATAATGTATGGCTGCGACGTGGGGCCGGA...            EXON       FEATURE     3     276           NaN\n",
       "7                        8  GTACCAGGGGCCACGGGGCGCCTCCCTGATCGCCTATAGATCTCCC...          INTRON       FEATURE     3     579           NaN\n",
       "8                        9  ACCCCCCCAAGACACATATGACCCACCACCCCATCTCTGACCATGA...            EXON       FEATURE     4     276           NaN\n",
       "9                       10  GTAAGGAGGGAGATGGGGGTGTCATGTCTCTTAGGGAAAGCAGGAG...          INTRON       FEATURE     4     102           NaN"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs['sequence_nodes.csv'].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cdsId:ID(CDS)</th>\n",
       "      <th>name</th>\n",
       "      <th>cdstype:LABEL</th>\n",
       "      <th>cds</th>\n",
       "      <th>protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ATGGCCGTCATGGCGCCCCGAACCCTCCTCCTGCTACTCTCGGGGG...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ATGGCCGTCATGGCGCCCCGAACCCTCCTCCTGCTACTCTCGGGGG...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>ATGGCCGTCATGGCGCCCCGAACCCTCCTCCTGCTACTCTCGGGGG...</td>\n",
       "      <td>MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...</td>\n",
       "      <td>SHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...</td>\n",
       "      <td>SHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...</td>\n",
       "      <td>SHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...</td>\n",
       "      <td>SHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...</td>\n",
       "      <td>SHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...</td>\n",
       "      <td>SHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...</td>\n",
       "      <td>SHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEP...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cdsId:ID(CDS) name cdstype:LABEL                                                cds                                            protein\n",
       "0              1  CDS           CDS  ATGGCCGTCATGGCGCCCCGAACCCTCCTCCTGCTACTCTCGGGGG...  MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...\n",
       "1              2  CDS           CDS  ATGGCCGTCATGGCGCCCCGAACCCTCCTCCTGCTACTCTCGGGGG...  MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...\n",
       "2              3  CDS           CDS  ATGGCCGTCATGGCGCCCCGAACCCTCCTCCTGCTACTCTCGGGGG...  MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRF...\n",
       "3              4  CDS           CDS  GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...  SHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEP...\n",
       "4              5  CDS           CDS  GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...  SHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEP...\n",
       "5              6  CDS           CDS  GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...  SHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEP...\n",
       "6              7  CDS           CDS  GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...  SHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEP...\n",
       "7              8  CDS           CDS  GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...  SHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEP...\n",
       "8              9  CDS           CDS  GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...  SHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEP...\n",
       "9             10  CDS           CDS  GCTCCCACTCCATGAGGTATTTCTTCACATCCGTGTCCCGGCCCGG...  SHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEP..."
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs['cds_nodes.csv'].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>:START_ID(SEQUENCE)</th>\n",
       "      <th>:END_ID(CDS)</th>\n",
       "      <th>:TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>36</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>38</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>45</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_CDS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   :START_ID(SEQUENCE)  :END_ID(CDS)    :TYPE\n",
       "0                    1             1  HAS_CDS\n",
       "1                   19             1  HAS_CDS\n",
       "2                   23             1  HAS_CDS\n",
       "3                   26             1  HAS_CDS\n",
       "4                   30             1  HAS_CDS\n",
       "5                   33             1  HAS_CDS\n",
       "6                   36             1  HAS_CDS\n",
       "7                   38             1  HAS_CDS\n",
       "8                   42             1  HAS_CDS\n",
       "9                   45             1  HAS_CDS"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs['cds_edges.csv'].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>:START_ID(ALLELE)</th>\n",
       "      <th>:END_ID(ALLELE)</th>\n",
       "      <th>imgt_release</th>\n",
       "      <th>:TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3360</td>\n",
       "      <td>HAS_GFE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>3360</td>\n",
       "      <td>HAS_GFE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>3360</td>\n",
       "      <td>HAS_GFE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>3360</td>\n",
       "      <td>HAS_GFE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>3360</td>\n",
       "      <td>HAS_GFE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>17</td>\n",
       "      <td>16</td>\n",
       "      <td>3360</td>\n",
       "      <td>HAS_GFE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>3360</td>\n",
       "      <td>HAS_GFE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>21</td>\n",
       "      <td>20</td>\n",
       "      <td>3360</td>\n",
       "      <td>HAS_GFE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>3360</td>\n",
       "      <td>HAS_GFE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>25</td>\n",
       "      <td>24</td>\n",
       "      <td>3360</td>\n",
       "      <td>HAS_GFE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   :START_ID(ALLELE)  :END_ID(ALLELE)  imgt_release    :TYPE\n",
       "0                  2                1          3360  HAS_GFE\n",
       "1                  8                7          3360  HAS_GFE\n",
       "2                 11               10          3360  HAS_GFE\n",
       "3                 13               12          3360  HAS_GFE\n",
       "4                 15               14          3360  HAS_GFE\n",
       "5                 17               16          3360  HAS_GFE\n",
       "6                 19               18          3360  HAS_GFE\n",
       "7                 21               20          3360  HAS_GFE\n",
       "8                 23               22          3360  HAS_GFE\n",
       "9                 25               24          3360  HAS_GFE"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs['gfe_edges.csv'].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>:START_ID(ALLELE)</th>\n",
       "      <th>:END_ID(SEQUENCE)</th>\n",
       "      <th>imgt_release</th>\n",
       "      <th>accession</th>\n",
       "      <th>:TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3360</td>\n",
       "      <td>0</td>\n",
       "      <td>HAS_SEQUENCE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3360</td>\n",
       "      <td>0</td>\n",
       "      <td>HAS_SEQUENCE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3360</td>\n",
       "      <td>2</td>\n",
       "      <td>HAS_FEATURE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3360</td>\n",
       "      <td>2</td>\n",
       "      <td>HAS_FEATURE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3360</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_FEATURE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3360</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_FEATURE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>3360</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_FEATURE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>3360</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_FEATURE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>3360</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_FEATURE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>3360</td>\n",
       "      <td>1</td>\n",
       "      <td>HAS_FEATURE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   :START_ID(ALLELE)  :END_ID(SEQUENCE)  imgt_release  accession         :TYPE\n",
       "0                  1                  1          3360          0  HAS_SEQUENCE\n",
       "1                  2                  1          3360          0  HAS_SEQUENCE\n",
       "2                  2                  2          3360          2   HAS_FEATURE\n",
       "3                  1                  2          3360          2   HAS_FEATURE\n",
       "4                  2                  3          3360          1   HAS_FEATURE\n",
       "5                  1                  3          3360          1   HAS_FEATURE\n",
       "6                  2                  4          3360          1   HAS_FEATURE\n",
       "7                  1                  4          3360          1   HAS_FEATURE\n",
       "8                  2                  5          3360          1   HAS_FEATURE\n",
       "9                  1                  5          3360          1   HAS_FEATURE"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs['seq_edges.csv'].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>:START_ID(ALLELE)</th>\n",
       "      <th>:END_ID(ALLELE)</th>\n",
       "      <th>imgtdb</th>\n",
       "      <th>:TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3360</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>3360</td>\n",
       "      <td>lg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>3360</td>\n",
       "      <td>lgx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>3360</td>\n",
       "      <td>2nd_FIELD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>3360</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>3360</td>\n",
       "      <td>lg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>3360</td>\n",
       "      <td>lgx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>3360</td>\n",
       "      <td>2nd_FIELD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>3360</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>3360</td>\n",
       "      <td>lg</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   :START_ID(ALLELE)  :END_ID(ALLELE)  imgtdb      :TYPE\n",
       "0                  3                2    3360          G\n",
       "1                  4                2    3360         lg\n",
       "2                  5                2    3360        lgx\n",
       "3                  6                2    3360  2nd_FIELD\n",
       "4                  3                8    3360          G\n",
       "5                  4                8    3360         lg\n",
       "6                  5                8    3360        lgx\n",
       "7                  9                8    3360  2nd_FIELD\n",
       "8                  3               11    3360          G\n",
       "9                  4               11    3360         lg"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs['group_edges.csv'].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
